src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs

```diff
@@ -0,0 +1,224 @@
+
+        public void ForEachScope<TState>(Action<object, TState> callback, TState state)
+        {
+            void Report(Scope current)
+            {
+                if (current == null)
+                {
+                    return;
+                }
+                Report(current.Parent);
+                callback(current.State, state);
+            }
+
+            if (_activityTrackingOption != ActivityTrackingOptions.None)
+            {
+                Activity activity = Activity.Current;
+                if (activity != null)
+                {
+                    const string propertyKey = "__ActivityLogScope__";
+
+                    ActivityLogScope activityLogScope = activity.GetCustomProperty(propertyKey) as ActivityLogScope;
+                    if (activityLogScope == null)
+                    {
+                        activityLogScope = new ActivityLogScope(activity, _activityTrackingOption);
+                        activity.SetCustomProperty(propertyKey, activityLogScope);
+                    }
+
+                    callback(activityLogScope, state);
+                }
+            }
+
+            Report(_currentScope.Value);
+        }
+
+        public IDisposable Push(object state)
+        {
+            var parent = _currentScope.Value;
+            var newScope = new Scope(this, state, parent);
+            _currentScope.Value = newScope;
+
+            return newScope;
+        }
+
+        private class Scope : IDisposable
+        {
+            private readonly LoggerFactoryScopeProvider _provider;
+            private bool _isDisposed;
+
+            internal Scope(LoggerFactoryScopeProvider provider, object state, Scope parent)
+            {
+                _provider = provider;
+                State = state;
+                Parent = parent;
+            }
+
+            public Scope Parent { get; }
+
+            public object State { get; }
+
+            public override string ToString()
+            {
+                return State?.ToString();
+            }
+
+            public void Dispose()
+            {
+                if (!_isDisposed)
+                {
+                    _provider._currentScope.Value = Parent;
+                    _isDisposed = true;
+                }
+            }
+        }
+
+        private class ActivityLogScope : IReadOnlyList<KeyValuePair<string, object>>
+        {
+            private readonly Activity _activity;
+            private readonly ActivityTrackingOptions _activityTrackingOption;
+
+            private string _cachedToString;
+
+            public int Count
+            {
+                get
+                {
+                    return 3;
+                }
+            }
+
+            public KeyValuePair<string, object> this[int index]
+            {
+                get
+                {
+                    if (index == 0)
```

Comment:
I think these are supposed to be based on the flags that were set.

Reply:
Thanks for the catch. This part missed when I added the enums. I have fixed it now.